packages/python/plotly/plotly/express/_core.py

```diff
@@ -754,6 +754,119 @@ def apply_default_cascade(args):
+    df_columns = (
+        args["data_frame"].columns if args.get("data_frame") is not None else None
+    )
+    group_attrs = ["symbol", "line_dash"]
+    for group_attr in group_attrs:
+        if group_attr in args:
+            attrables += [group_attr]
+
+    # Loop over possible arguments
+    for field_name in attrables:
+        argument_list = (
+            [args.get(field_name)]
+            if field_name not in array_attrables
+            else args.get(field_name)
+        )
+        # argument not specified, continue
+        if argument_list is None or argument_list is [None]:
+            continue
+        # Argument name: field_name if the argument is a list
+        # Else we give names like ["hover_data_0, hover_data_1"] etc.
+        field_list = (
+            [field_name]
+            if field_name not in array_attrables
+            else [field_name + "_" + str(i) for i in range(len(argument_list))]
+        )
+        # argument_list and field_list ready, iterate over them
+        for i, (argument, field) in enumerate(zip(argument_list, field_list)):
+            if argument is None:
+                continue
+            elif isinstance(argument, str):  # just a column name
+                if not isinstance(args.get("data_frame"), pd.DataFrame):
+                    raise ValueError(
+                        "String arguments are only possible when a DataFrame"
+                        "is provided in the `data_frame` argument."
+                    )
+                # Check validity of column name
+                try:
+                    df[argument] = args["data_frame"][argument]
+                    continue
+                except KeyError:
+                    raise ValueError(
+                        "Value of '%s' is not the name of a column in 'data_frame'. "
+                        "Expected one of %s but received: %s"
+                        % (field, str(list(df_columns)), argument)
+                    )
+            # Case of index
+            elif isinstance(argument, pd.core.indexes.range.RangeIndex):
+                col_name = argument.name if argument.name else "index"
+                try:
+                    df.insert(0, col_name, argument)
+                except ValueError:  # if col named index already exists, replace
+                    df[col_name] = argument
+            # Case of numpy array or df column
+            else:
+                try:
+                    col_name = argument.name  # pandas df
+                    if (
+                        args.get("data_frame") is not None
+                        and col_name in args["data_frame"]
+                    ):
+                        # If the name exists but the values have changed
+                        # we do not want to keep the name, revert to field
+                        col_name = (
+                            col_name
+                            if args["data_frame"][col_name].equals(argument)
```

Comment:
what's the tradeoff between using `.equals()` vs `is` here? the former is slower but a bit more permissive in that if it's a copy with the same name we accept it?

Reply:
yes this was my idea. But maybe the flexibility gain is not worth the performance cost, I'm also happy to change to `is`.

Reply:
Maybe we can switch to `is` for maximum strictness/performance, to start, and see if people complain/notice?

Reply:
done